docs/src/adr/015adr-trace.md

```diff
@@ -0,0 +1,403 @@
+      - quite hard to parse automatically.
+
+ 1. Use the JSON serialization as in [ADR005][]:
+
+    - Pros:
+
+      - easy to parse automatically.
+
+    - Cons:
+
+      - almost impossible to read.
+      - requires the knowledge of Apalache IR.
+
+## Solution
+
+In this ADR, we propose a very simple format that represents executions of
+state machines that follows the concepts of TLA+ and yet avoids complexity of
+TLA+. It is so simple that we call it *"Informal Trace Format"* (ITF).
+(Obviously, it is formal enough to be machine-readable).  By convention, the
+files in this format should end with the extension `.itf.json`.
+
+### The ITF Format
+
+**Trace object**. A trace in ITF is a JSON object:
+
+```js
+{
+  "#meta": <optional object>,
+  "params": <optional array of strings>,
+  "vars":  <array of strings>,
+  "states": <array of states>,
+  "loop": <optional int>
+}
+```
+
+The field `#meta` is an arbitrary JSON object, whose purpose is to provide
+the reader with additional comments about the trace. For example, it may look
+like:
+
+```js
+  "#meta": {
+    "description": "Generated by Apalache",
+    "source": "MissionariesAndCannibalsTyped.tla"
+  }
+```
+
+The optional field `params` is an array of names that must be set in the
+initial state (if there are any parameters). The parameters play the same role
+as `CONSTANTS` in TLA+. For example, the field may look like:
+
+```js
+  "params": [ "Missionaries", "Cannibals" ]
+```
+
+The field `vars` is an array of names that must be set in every state.
+For example, the field may look like:
+
+```js
+  "vars": [ "bank_of_boat", "who_is_on_bank" ]
+```
+
+The field `states` is an array of state objects (see below). For example,
+the field may look like:
+
+```js
+  "states": [ <state0>, <state1>, <state2> ]
+```
+
+The optional field `loop` specifies the index of the state (in the array of
+states) that starts the loop. The loop ends in the last state. For example,
+the field may look like:
+
+```js
+  "loop": 1
+```
+
+**State object**. A state is a JSON object:
+
+```js
+  {
+    "#meta": <optional object>,
+    "<var1>": <expr>,
+    ...
+    "<varN>": <expr>
+  }
+```
+
+As in the trace object, the field `#meta` may be an arbitrary object.
+Different tools may use this object to write their metadata into this object.
+
+The names `<var1>, ..., <varN>` are the names of the variables that are
+specified in the field `vars`. Each state must define a value for every specified variable. The syntax of `<expr>` is specified below. 
+
+**Expressions.** As usual, expressions are inductively defined. An expression
+`<expr>` is one of the following:
+
+ 1. A JSON string literal, e.g., `"hello"`. TLA+ strings are written as strings
+    in this format.
+
+ 1. A JSON integer, e.g., 123. According to [RFC7159][], JSON integers
+    must be in the range: `[-(2**53)+1, (2**53)-1]`. Integers in this
+    range *may be* written as JSON integers.
```

Comment:
As a potential use of this encoding, I don't like that some values just the literal JSON value, while others are inferred from a special encoding. I suggest instead just making every value of the form `{"#type": <type-name>, "#value": <value>}`.

Otherwise people writer deserializers have to look at each object and inspect it first to see if it's one of these special objects, by getting all keys, then checking for the presence of the special key, or if it's just a normal object (which is actually supposed to be interpreted as a TLA+ record). This ends up introducing two different dimensions of ambiguity: some json values are puns for certain TLA+ values, and some json values are actually special instructions that are meant to indicate an alternative interpretation.

I think you can make the format much more regular if instead every value is tagged, then the function to dispatch for deserialization can just be something like

```scala
def deser(o: Json): Tla = o.type match {
  case "#string" => Some(o.value)
  case "#int" => Option(o.value.toInt)
  case "#rec" => convertRecord(o.value)
  // ...
}
```

My main claim here is that requiring tags for some values but not others, and putting these tags each as their own dictionary key instead of assigning a single key to pull out tags, actually doesn't provide any simplification, it just makes things more irregular.

Reply:
I had the same feeling when writing this document. Actually, I wanted to keep JSON values only, but that seemed to be too much loss of information. We could make the values more regular; it's easy. However, I think that the priorities are different here:

1. Some of these traces may be written by hand, when people are developing their own tools. In this case, people would probably limit themselves to arrays and records.
2. When I am looking at the testing frameworks that we do for audits, the structure of the counterexamples is fixed, so we already know the shape of the states.

Reply:
I agree with Shon here for what it's worth. The ADR looks good a whole